src/Symfony/Component/Message/HandlerLocator.php

```diff
@@ -0,0 +1,56 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Message;
+
+use Symfony\Component\Message\Exception\NoHandlerForMessageException;
+use Symfony\Component\Message\Handler\ChainHandler;
+
+/**
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ */
+class HandlerLocator implements HandlerLocatorInterface
+{
+    /**
+     * Maps a message (its class) to a given handler.
+     *
+     * @var array
+     */
+    private $messageToHandlerMapping;
+
+    public function __construct(array $messageToHandlerMapping = array())
+    {
+        $this->messageToHandlerMapping = $messageToHandlerMapping;
+    }
+
+    public function resolve($message): callable
+    {
+        $messageKey = get_class($message);
+
+        if (!isset($this->messageToHandlerMapping[$messageKey])) {
+            throw new NoHandlerForMessageException(sprintf('No handler for message "%s".', $messageKey));
+        }
+
+        $handler = $this->messageToHandlerMapping[$messageKey];
+        if ($this->isCollectionOfHandlers($handler)) {
```

Comment:
we already handle this at the wiring level (`MessagePass`), meaning the contract could be simply `key => handler`.